In [2]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import sklearn
import shap

print("Environment is ready!")
print("Pandas:", pd.__version__)
print("NumPy:", np.__version__)
print("Scikit-learn:", sklearn.__version__)

Environment is ready!
Pandas: 3.0.5
NumPy: 2.5.2
Scikit-learn: 1.9.0


In [4]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

import warnings
warnings.filterwarnings("ignore")

pd.set_option("display.max_columns", None)
pd.set_option("display.max_rows", 100)

print("Libraries imported successfully.")

Libraries imported successfully.


In [9]:
df = pd.read_csv("../data/raw/tech_advertising_campaigns_dataset.csv")
print("Dataset loaded successfully.")

Dataset loaded successfully.


In [10]:
df.head()

,campaign_id,campaign_objective,platform,ad_placement,device_type,operating_system,creative_format,creative_size,ad_copy_length,has_call_to_action,creative_emotion,creative_age_days,target_audience_age,target_audience_gender,audience_interest_category,income_bracket,purchase_intent_score,retargeting_flag,start_date,quarter,day_of_week,hour_of_day,campaign_day,quality_score,actual_cpc,impressions,clicks,conversions,ad_spend,revenue,bounce_rate,avg_session_duration_seconds,pages_per_session,industry_vertical,budget_tier,CTR,CPC,conversion_rate,CPA,ROAS,profit
0,CAMP_00001,Lead Generation,Facebook,Search,Mobile,Android,Text,728x90,Short,False,Curiosity,72,45-54,Female,Shoppers,<$50K,Medium,False,06-03-2024,1,Wednesday,18,34,5,2.23,66329,402,2,896.46,289.67,67.75,70,2.00,E-commerce,High,0.606,2.23,0.498,448.23,0.32,-606.79
1,CAMP_00002,Engagement,Facebook,Feed,Mobile,iOS,Image,320x50,Long,True,Neutral,62,65+,Female,Business Professionals,$50K-$100K,Medium,False,26-01-2024,1,Friday,6,22,5,2.48,50094,467,5,1158.16,1728.20,63.73,105,2.74,Finance,Medium,0.932,2.48,1.071,231.63,1.49,570.04
2,CAMP_00003,Conversions,Google Ads,Feed,Tablet,iOS,Video,1920x1080,Short,False,Urgency,79,65+,Male,Tech Enthusiasts,<$50K,High,False,15-05-2025,2,Thursday,1,32,4,3.88,10842,112,6,434.56,2903.26,31.93,149,3.97,Healthcare,Low,1.033,3.88,5.357,72.43,6.68,2468.70
3,CAMP_00004,Conversions,LinkedIn,Search,Desktop,iOS,Carousel,1920x1080,Short,False,Joy,57,35-44,All,Shoppers,<$50K,Low,False,21-07-2024,3,Sunday,7,32,4,8.29,7820,123,3,1019.67,502.28,65.52,147,2.49,Education,Medium,1.573,8.29,2.439,339.89,0.49,-517.39
4,CAMP_00005,Brand Awareness,Facebook,Stories,Mobile,iOS,Image,1920x1080,Short,False,Joy,17,25-34,Female,Students,$100K-$200K,Low,False,09-03-2025,1,Sunday,17,52,7,1.41,21436,302,0,425.82,0.00,68.95,81,3.04,SaaS,Low,1.409,1.41,0.000,0.00,0.00,-425.82


In [11]:
print("Dataset Shape:", df.shape)
print("Number of Rows:", df.shape[0])
print("Number of Columns:", df.shape[1])

Dataset Shape: (10000, 41)
Number of Rows: 10000
Number of Columns: 41


In [12]:
print("Column Names:\n")

for i, column in enumerate(df.columns, start=1):
    print(f"{i:02d}. {column}")

Column Names:

01. campaign_id
02. campaign_objective
03. platform
04. ad_placement
05. device_type
06. operating_system
07. creative_format
08. creative_size
09. ad_copy_length
10. has_call_to_action
11. creative_emotion
12. creative_age_days
13. target_audience_age
14. target_audience_gender
15. audience_interest_category
16. income_bracket
17. purchase_intent_score
18. retargeting_flag
19. start_date
20. quarter
21. day_of_week
22. hour_of_day
23. campaign_day
24. quality_score
25. actual_cpc
26. impressions
27. clicks
28. conversions
29. ad_spend
30. revenue
31. bounce_rate
32. avg_session_duration_seconds
33. pages_per_session
34. industry_vertical
35. budget_tier
36. CTR
37. CPC
38. conversion_rate
39. CPA
40. ROAS
41. profit


In [13]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 41 columns):
 #   Column                        Non-Null Count  Dtype  
---  ------                        --------------  -----  
 0   campaign_id                   10000 non-null  str    
 1   campaign_objective            10000 non-null  str    
 2   platform                      10000 non-null  str    
 3   ad_placement                  10000 non-null  str    
 4   device_type                   10000 non-null  str    
 5   operating_system              10000 non-null  str    
 6   creative_format               10000 non-null  str    
 7   creative_size                 10000 non-null  str    
 8   ad_copy_length                10000 non-null  str    
 9   has_call_to_action            10000 non-null  bool   
 10  creative_emotion              10000 non-null  str    
 11  creative_age_days             10000 non-null  int64  
 12  target_audience_age           10000 non-null  str    
 13  target_audien

In [14]:
df.describe()

,creative_age_days,quarter,hour_of_day,campaign_day,quality_score,actual_cpc,impressions,clicks,conversions,ad_spend,revenue,bounce_rate,avg_session_duration_seconds,pages_per_session,CTR,CPC,conversion_rate,CPA,ROAS,profit
count,10000.0000,10000.000000,10000.000000,10000.000000,10000.000000,10000.000000,10000.000000,10000.000000,10000.000000,10000.000000,1.000000e+04,10000.000000,10000.000000,10000.000000,10000.000000,10000.000000,10000.000000,10000.000000,10000.000000,1.000000e+04
mean,45.3743,2.419100,11.482500,45.429700,5.527400,3.290336,70539.223000,1526.436100,65.568800,4345.566179,2.841577e+04,52.528500,100.956600,3.113223,2.314340,3.290336,3.644033,190.895894,7.524918,2.407020e+04
std,26.1711,1.132424,6.932308,26.071107,1.553541,2.453852,91339.733246,2412.164582,183.404876,6917.481037,8.928461e+04,17.706165,46.541295,1.369884,1.573316,2.453852,4.354338,295.614462,14.985237,8.641610e+04
min,1.0000,1.000000,0.000000,1.000000,1.000000,0.300000,5000.000000,10.000000,0.000000,8.690000,0.000000e+00,10.000000,10.000000,1.000000,0.100000,0.300000,0.000000,0.000000,0.000000,-3.693984e+04
25%,22.0000,1.000000,5.000000,23.000000,4.000000,1.510000,16958.500000,290.000000,5.000000,666.307500,1.608910e+03,40.087500,68.000000,2.160000,1.172000,1.510000,1.005000,40.640000,1.197500,2.094800e+02
50%,45.0000,2.000000,12.000000,45.000000,6.000000,2.430000,37469.500000,718.000000,16.000000,1895.670000,6.099240e+03,53.580000,100.000000,2.830000,1.920000,2.430000,2.165500,99.460000,3.150000,3.302535e+03
75%,68.0000,3.000000,18.000000,68.000000,7.000000,4.530000,83213.750000,1746.000000,55.000000,4996.160000,2.216582e+04,65.472500,132.000000,3.740000,3.017250,4.530000,4.493250,225.457500,7.950000,1.664188e+04
max,90.0000,4.000000,23.000000,90.000000,10.000000,16.030000,500000.000000,40000.000000,5384.000000,49999.680000,2.800397e+06,90.000000,277.000000,8.510000,8.000000,16.030000,25.000000,6379.940000,487.300000,2.750398e+06


In [16]:
missing_summary = pd.DataFrame({
    "Missing_Count": df.isnull().sum(),
    "Missing_Percentage": (df.isnull().sum() / len(df)) * 100
})

missing_summary = missing_summary.sort_values(
    "Missing_Count",
    ascending=False
)

missing_summary

,Missing_Count,Missing_Percentage
campaign_id,0,0.0
campaign_objective,0,0.0
platform,0,0.0
ad_placement,0,0.0
device_type,0,0.0
operating_system,0,0.0
creative_format,0,0.0
creative_size,0,0.0
ad_copy_length,0,0.0
has_call_to_action,0,0.0


In [17]:
print("Duplicate rows:", df.duplicated().sum())

Duplicate rows: 0


In [18]:
numerical_columns = df.select_dtypes(
    include=np.number
).columns.tolist()

categorical_columns = df.select_dtypes(
    exclude=np.number
).columns.tolist()

print("Number of numerical variables:", len(numerical_columns))
print("Number of categorical/non-numerical variables:", len(categorical_columns))

Number of numerical variables: 20
Number of categorical/non-numerical variables: 21


In [19]:
for column in categorical_columns:
    print(f"\n{'=' * 70}")
    print(f"Column: {column}")
    print(f"Unique Values: {df[column].nunique()}")
    print(df[column].unique()[:20])


Column: campaign_id
Unique Values: 10000
<StringArray>
['CAMP_00001', 'CAMP_00002', 'CAMP_00003', 'CAMP_00004', 'CAMP_00005',
 'CAMP_00006', 'CAMP_00007', 'CAMP_00008', 'CAMP_00009', 'CAMP_00010',
 'CAMP_00011', 'CAMP_00012', 'CAMP_00013', 'CAMP_00014', 'CAMP_00015',
 'CAMP_00016', 'CAMP_00017', 'CAMP_00018', 'CAMP_00019', 'CAMP_00020']
Length: 20, dtype: str

Column: campaign_objective
Unique Values: 5
<StringArray>
['Lead Generation',      'Engagement',     'Conversions', 'Brand Awareness',
    'App Installs']
Length: 5, dtype: str

Column: platform
Unique Values: 6
<StringArray>
['Facebook', 'Google Ads', 'LinkedIn', 'Twitter', 'TikTok', 'Instagram']
Length: 6, dtype: str

Column: ad_placement
Unique Values: 6
<StringArray>
['Search', 'Feed', 'Stories', 'Sidebar', 'In-Stream Video', 'Display Network']
Length: 6, dtype: str

Column: device_type
Unique Values: 3
<StringArray>
['Mobile', 'Tablet', 'Desktop']
Length: 3, dtype: str

Column: operating_system
Unique Values: 5
<StringArray

First data-quality check

In [21]:
important_columns = [
    "impressions",
    "clicks",
    "conversions",
    "ad_spend",
    "revenue",
    "CTR",
    "conversion_rate",
    "ROAS",
    "profit"
]

available_columns = [
    column for column in important_columns
    if column in df.columns
]

df[available_columns].describe().T

,count,mean,std,min,25%,50%,75%,max
impressions,10000.0,70539.223000,91339.733246,5000.00,16958.5000,37469.5000,83213.75000,500000.00
clicks,10000.0,1526.436100,2412.164582,10.00,290.0000,718.0000,1746.00000,40000.00
conversions,10000.0,65.568800,183.404876,0.00,5.0000,16.0000,55.00000,5384.00
ad_spend,10000.0,4345.566179,6917.481037,8.69,666.3075,1895.6700,4996.16000,49999.68
revenue,10000.0,28415.770647,89284.609131,0.00,1608.9100,6099.2400,22165.82500,2800397.43
CTR,10000.0,2.314340,1.573316,0.10,1.1720,1.9200,3.01725,8.00
conversion_rate,10000.0,3.644033,4.354338,0.00,1.0050,2.1655,4.49325,25.00
ROAS,10000.0,7.524918,14.985237,0.00,1.1975,3.1500,7.95000,487.30
profit,10000.0,24070.204468,86416.098554,-36939.84,209.4800,3302.5350,16641.87750,2750397.93


In [22]:
phase1_summary = pd.DataFrame({
    "Metric": [
        "Rows",
        "Columns",
        "Duplicate Rows",
        "Total Missing Values",
        "Numerical Variables",
        "Categorical / Non-Numerical Variables"
    ],
    "Value": [
        df.shape[0],
        df.shape[1],
        df.duplicated().sum(),
        df.isnull().sum().sum(),
        len(numerical_columns),
        len(categorical_columns)
    ]
})

phase1_summary

,Metric,Value
0,Rows,10000
1,Columns,41
2,Duplicate Rows,0
3,Total Missing Values,0
4,Numerical Variables,20
5,Categorical / Non-Numerical Variables,21


In [23]:
# Calculate KPI values from the raw metrics

df["CTR_calculated"] = (df["clicks"] / df["impressions"]) * 100

df["conversion_rate_calculated"] = (
    df["conversions"] / df["clicks"]
) * 100

df["ROAS_calculated"] = (
    df["revenue"] / df["ad_spend"]
)

df["profit_calculated"] = (
    df["revenue"] - df["ad_spend"]
)

print("KPI calculations completed.")

KPI calculations completed.


In [24]:
kpi_comparison = pd.DataFrame({
    "CTR_difference": df["CTR"] - df["CTR_calculated"],
    "Conversion_Rate_difference":
        df["conversion_rate"] - df["conversion_rate_calculated"],
    "ROAS_difference":
        df["ROAS"] - df["ROAS_calculated"],
    "Profit_difference":
        df["profit"] - df["profit_calculated"]
})

kpi_comparison.describe().T

,count,mean,std,min,25%,50%,75%,max
CTR_difference,10000.0,3.370958e-06,2.836611e-04,-4.999517e-04,-0.000235,0.0,0.000247,4.999710e-04
Conversion_Rate_difference,10000.0,-1.345144e-06,2.796086e-04,-5.000000e-04,-0.000238,0.0,0.000230,5.000000e-04
ROAS_difference,10000.0,-2.566521e-05,2.809777e-03,-4.999824e-03,-0.002425,0.0,0.002353,5.000000e-03
Profit_difference,10000.0,1.176957e-14,7.165006e-12,-2.328306e-10,0.000000,0.0,0.000000,4.656613e-10


In [27]:
temporary_columns = [
    "CTR_calculated",
    "conversion_rate_calculated",
    "ROAS_calculated",
    "profit_calculated"
]

df.drop(columns=temporary_columns, inplace=True)

print("Temporary KPI calculation columns removed.")
print("Current dataset shape:", df.shape)

Temporary KPI calculation columns removed.
Current dataset shape: (10000, 41)


In [28]:
df.select_dtypes(include="object").nunique().sort_values()

device_type                       3
purchase_intent_score             3
target_audience_gender            3
ad_copy_length                    3
budget_tier                       3
income_bracket                    4
operating_system                  5
campaign_objective                5
target_audience_age               6
creative_size                     6
ad_placement                      6
creative_format                   6
audience_interest_category        6
creative_emotion                  6
industry_vertical                 6
platform                          6
day_of_week                       7
start_date                      761
campaign_id                   10000
dtype: int64

In [29]:
df.select_dtypes(include=np.number).nunique().sort_values()

quarter                            4
quality_score                     10
hour_of_day                       24
creative_age_days                 90
campaign_day                      90
avg_session_duration_seconds     238
conversions                      619
pages_per_session                651
actual_cpc                      1084
CPC                             1084
ROAS                            2368
clicks                          3499
CTR                             4181
conversion_rate                 5085
bounce_rate                     5155
CPA                             8028
impressions                     9037
ad_spend                        9301
revenue                         9385
profit                          9953
dtype: int64

In [30]:
df["CPC_calculated"] = df["ad_spend"] / df["clicks"]

df["CPA_calculated"] = df["ad_spend"] / df["conversions"]

df["actual_cpc_difference"] = (
    df["actual_cpc"] - df["CPC_calculated"]
)

df["CPC_difference"] = (
    df["CPC"] - df["CPC_calculated"]
)

df["CPA_difference"] = (
    df["CPA"] - df["CPA_calculated"]
)

df[
    [
        "actual_cpc_difference",
        "CPC_difference",
        "CPA_difference"
    ]
].describe().T

,count,mean,std,min,25%,50%,75%,max
actual_cpc_difference,10000.0,-1.071365e-18,3.343760e-16,-1.776357e-15,0.000000,0.0,0.000000,3.552714e-15
CPC_difference,10000.0,-1.071365e-18,3.343760e-16,-1.776357e-15,0.000000,0.0,0.000000,3.552714e-15
CPA_difference,10000.0,-inf,NaN,-inf,-0.002692,0.0,0.001861,5.000000e-03


In [32]:
temporary_columns = [
    "CPC_calculated",
    "CPA_calculated",
    "actual_cpc_difference",
    "CPC_difference",
    "CPA_difference"
]

df.drop(
    columns=temporary_columns,
    inplace=True,
    errors="ignore"
)

print("Temporary validation columns removed.")
print("Dataset shape:", df.shape)

Temporary validation columns removed.
Dataset shape: (10000, 41)


In [33]:
print("Clicks greater than impressions:",
      (df["clicks"] > df["impressions"]).sum())

print("Conversions greater than clicks:",
      (df["conversions"] > df["clicks"]).sum())

print("Negative impressions:",
      (df["impressions"] < 0).sum())

print("Negative clicks:",
      (df["clicks"] < 0).sum())

print("Negative conversions:",
      (df["conversions"] < 0).sum())

print("Negative ad spend:",
      (df["ad_spend"] < 0).sum())

print("Negative revenue:",
      (df["revenue"] < 0).sum())

Clicks greater than impressions: 0
Conversions greater than clicks: 0
Negative impressions: 0
Negative clicks: 0
Negative conversions: 0
Negative ad spend: 0
Negative revenue: 0


In [34]:
print("\nTarget variable ranges:")

for column in ["CTR", "conversion_rate", "ROAS", "profit"]:
    print(
        f"{column}: "
        f"min={df[column].min()}, "
        f"max={df[column].max()}"
    )


Target variable ranges:
CTR: min=0.1, max=8.0
conversion_rate: min=0.0, max=25.0
ROAS: min=0.0, max=487.3
profit: min=-36939.84, max=2750397.93


In [35]:
numeric_kpis = [
    "impressions",
    "clicks",
    "conversions",
    "ad_spend",
    "revenue",
    "CTR",
    "conversion_rate",
    "CPC",
    "CPA",
    "ROAS",
    "profit"
]

outlier_summary = []

for column in numeric_kpis:
    Q1 = df[column].quantile(0.25)
    Q3 = df[column].quantile(0.75)
    IQR = Q3 - Q1

    lower_bound = Q1 - 1.5 * IQR
    upper_bound = Q3 + 1.5 * IQR

    outliers = (
        (df[column] < lower_bound) |
        (df[column] > upper_bound)
    ).sum()

    outlier_summary.append({
        "Variable": column,
        "Q1": Q1,
        "Q3": Q3,
        "IQR": IQR,
        "Lower_Bound": lower_bound,
        "Upper_Bound": upper_bound,
        "Outlier_Count": outliers,
        "Outlier_Percentage": (outliers / len(df)) * 100
    })

outlier_summary = pd.DataFrame(outlier_summary)

outlier_summary.sort_values(
    "Outlier_Percentage",
    ascending=False
)

,Variable,Q1,Q3,IQR,Lower_Bound,Upper_Bound,Outlier_Count,Outlier_Percentage
10,profit,209.4800,16641.87750,16432.39750,-24439.116250,41290.473750,1306,13.06
2,conversions,5.0000,55.00000,50.00000,-70.000000,130.000000,1206,12.06
4,revenue,1608.9100,22165.82500,20556.91500,-29226.462500,53001.197500,1182,11.82
9,ROAS,1.1975,7.95000,6.75250,-8.931250,18.078750,1016,10.16
3,ad_spend,666.3075,4996.16000,4329.85250,-5828.471250,11490.938750,955,9.55
0,impressions,16958.5000,83213.75000,66255.25000,-82424.375000,182596.625000,901,9.01
1,clicks,290.0000,1746.00000,1456.00000,-1894.000000,3930.000000,894,8.94
8,CPA,40.6400,225.45750,184.81750,-236.586250,502.683750,838,8.38
6,conversion_rate,1.0050,4.49325,3.48825,-4.227375,9.725625,815,8.15
5,CTR,1.1720,3.01725,1.84525,-1.595875,5.785125,429,4.29


# Phase 1 — Data Understanding: Conclusions

## 1. Dataset Overview

The advertising campaign dataset contains **10,000 observations and 41 variables**. The dataset consists of **20 numerical variables and 21 categorical/non-numerical variables**.

The variables cover multiple aspects of digital advertising campaigns, including campaign characteristics, platforms, creative attributes, audience characteristics, campaign quality and cost, performance metrics, financial outcomes, user engagement, and industry information.

---

## 2. Data Quality Assessment

The initial data-quality assessment produced the following results:

- **Total observations:** 10,000
- **Total variables:** 41
- **Missing values:** 0
- **Duplicate records:** 0
- **Numerical variables:** 20
- **Categorical/non-numerical variables:** 21

Therefore, the dataset contains no missing values or duplicate records and is structurally suitable for further analysis.

Logical consistency checks also showed:

- No cases where clicks exceeded impressions.
- No cases where conversions exceeded clicks.
- No negative impressions.
- No negative clicks.
- No negative conversions.
- No negative advertising spend.
- No negative revenue.

These results indicate that the major campaign volume and financial variables satisfy the basic logical constraints expected from advertising data.

---

## 3. Variable Classification

The variables were examined based on their data types and business meaning.

Important variable categories identified include:

### Identifier
- `campaign_id`

`campaign_id` contains 10,000 unique values and functions as a unique identifier. It will not be used as a predictive feature.

### Date Variable
- `start_date`

`start_date` represents the campaign start date and will require appropriate date-time processing during preprocessing and feature engineering.

### Nominal Categorical Variables

Examples include:

- `campaign_objective`
- `platform`
- `ad_placement`
- `device_type`
- `operating_system`
- `creative_format`
- `creative_size`
- `creative_emotion`
- `target_audience_age`
- `target_audience_gender`
- `audience_interest_category`
- `income_bracket`
- `industry_vertical`

### Ordinal Variables

The following variables contain ordered categories or scores:

- `purchase_intent_score` — Low, Medium, High
- `ad_copy_length` — Short, Medium, Long
- `budget_tier` — Low, Medium, High
- `quality_score` — 1 to 10

### Temporal and Duration Variables

- `quarter`
- `hour_of_day`
- `campaign_day`
- `creative_age_days`

These variables will be considered carefully during feature engineering because some have temporal or ordered characteristics.

---

## 4. Target Variables

The project focuses on four major campaign performance targets:

1. **CTR (Click-Through Rate)**
2. **Conversion Rate**
3. **ROAS (Return on Ad Spend)**
4. **Profit**

These targets will be used in the predictive modeling stage to estimate campaign performance and support marketing optimization decisions.

---

## 5. KPI Validation

The major campaign KPIs were validated against their underlying campaign metrics.

The following relationships were confirmed:

- **CTR = (Clicks / Impressions) × 100**
- **Conversion Rate = (Conversions / Clicks) × 100**
- **ROAS = Revenue / Ad Spend**
- **Profit = Revenue − Ad Spend**

The differences between the stored KPI values and the calculated values were extremely small, indicating that the stored KPI values are consistent with their underlying metrics, with minor differences attributable to rounding.

CPC and Actual CPC were also found to be derived from advertising spend and clicks.

---

## 6. Zero-Conversion Consideration

The dataset contains campaigns with zero conversions. Therefore, metrics such as CPA, which depend on the number of conversions, require special consideration.

For campaigns with zero conversions:

**CPA = Ad Spend / Conversions**

cannot be calculated using ordinary division because the denominator is zero.

This issue will be handled appropriately during the data preprocessing and feature engineering stages.

---

## 7. Distribution and Outlier Findings

Initial descriptive statistics indicate that several campaign performance and financial variables are highly skewed.

IQR-based outlier screening identified statistical outliers in several variables:

| Variable | Outlier Percentage |
|----------|-------------------:|
| Profit | 13.06% |
| Conversions | 12.06% |
| Revenue | 11.82% |
| ROAS | 10.16% |
| Ad Spend | 9.55% |
| Impressions | 9.01% |
| Clicks | 8.94% |
| CPA | 8.38% |
| Conversion Rate | 8.15% |
| CTR | 4.29% |
| CPC | 3.36% |

These observations will **not be removed automatically**.

In a marketing context, extreme values may represent legitimate high-performing or low-performing campaigns rather than data-entry errors. Therefore, the distributions and extreme observations will be investigated further during Exploratory Data Analysis.

---

## 8. Target Variable Ranges

The observed ranges of the four major target variables are:

| Target Variable | Minimum | Maximum |
|-----------------|--------:|--------:|
| CTR | 0.1 | 8.0 |
| Conversion Rate | 0.0 | 25.0 |
| ROAS | 0.0 | 487.3 |
| Profit | -36,939.84 | 2,750,397.93 |

The presence of negative profit values indicates that some campaigns generated financial losses, which is important for the campaign optimization component of the project.

The very high maximum ROAS and revenue/profit values will require further distributional investigation during EDA.

---

## 9. Predictive Modeling Considerations

The KPI validation revealed an important modeling consideration.

Since CTR, Conversion Rate, ROAS, and Profit are mathematically derived from other campaign metrics, directly using their underlying components as predictors could result in **target leakage or trivial prediction relationships**.

For example:

- Using clicks and impressions to predict CTR could allow the model to reproduce the CTR formula.
- Using conversions and clicks to predict Conversion Rate could produce a similar issue.
- Using revenue and ad spend to predict ROAS would directly expose the components of the target.
- Using revenue and ad spend to predict Profit would directly expose the profit calculation.

Therefore, predictive models should primarily focus on campaign characteristics, audience attributes, platform information, creative attributes, quality-related variables, and other information that can reasonably be available independently of the final campaign outcome.

The final feature-selection strategy will be determined during **Feature Engineering and Predictive Modeling**.

---

## 10. Overall Phase 1 Conclusion

The dataset is structurally suitable for the Marketing Intelligence for Campaign Optimization project.

The analysis established that:

- The dataset contains 10,000 observations and 41 variables.
- There are no missing values or duplicate records.
- Major campaign metrics satisfy basic logical consistency checks.
- The four main prediction targets are available and internally consistent.
- Several variables require special treatment based on their business meaning.
- Several performance and financial variables exhibit substantial statistical outliers and skewness.
- Extreme observations will be investigated rather than automatically removed.
- Potential target leakage has been identified and will be considered during feature engineering and model development.

The findings from this phase provide the foundation for the next stage of the project:

> **Phase 2 — Data Preprocessing**

In Phase 2, the dataset will be prepared for analysis and machine learning through appropriate data-type conversion, categorical encoding, handling of derived variables, treatment of special values, and preparation of modeling datasets.